src/eui/Eui64.cc

```diff
@@ -75,22 +80,99 @@ Eui::Eui64::lookupSlaac(const Ip::Address &c)
 
 // return binary representation of the EUI
 bool
-Eui::Eui64::lookupNdp(const Ip::Address &/*c*/)
+Eui::Eui64::lookupNdp(const Ip::Address &c)
 {
-#if 0 /* no actual lookup coded yet */
-
-    /* no OS yet supported for NDP protocol lookup */
+    bool success = false;
+
+#if _SQUID_LINUX_
+    int rtnetlink_sock;
+    struct {
+        struct nlmsghdr hdr;
+        struct ndmsg ndm;
+    } req;
+    struct sockaddr_nl sa;
+    struct iovec iov;
+    struct msghdr msg;
+    char buf[16384];
+    ssize_t buff_fill;
+    struct nlmsghdr * nl_reply_hdr;
+    struct nlmsghdr * nl_reply_msg;
+    struct rtattr * rt_reply_attr;
+    size_t rt_reply_len;
+    bool found_ip = 0;
+    bool done = 0;
+    struct in6_addr ip;
+    void * mac_ptr;
+
+    c.getInAddr(ip);
+
+    // Prepare rtnetlink request
+    req.hdr.nlmsg_len = NLMSG_LENGTH(sizeof(struct ndmsg));
+    req.hdr.nlmsg_type = RTM_GETNEIGH;
+    req.hdr.nlmsg_flags = NLM_F_REQUEST | NLM_F_DUMP;
+    req.hdr.nlmsg_seq = 0;
+    req.hdr.nlmsg_pid = 0;
+    req.ndm.ndm_family = AF_INET6;
+    req.ndm.ndm_state = 0;
+
+    iov.iov_base = &req;
+    iov.iov_len = req.hdr.nlmsg_len;
+
+    msg.msg_name = &sa;
+    msg.msg_namelen = sizeof(sa);
+    msg.msg_iov = &iov;
+    msg.msg_iovlen = 1;
+    msg.msg_control = NULL;
+    msg.msg_controllen = 0;
+    msg.msg_flags = 0;
+
+    // Send rtnetlink request
+    rtnetlink_sock = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+    if (rtnetlink_sock >= 0) {
+        memset(&sa, 0, sizeof(sa));
+        sa.nl_family = AF_NETLINK;
+        if (sendmsg(rtnetlink_sock, &msg, 0) >= 0) {
+            // Loop through all the responses
+            while ((! done) && ((buff_fill = recv(rtnetlink_sock, buf, sizeof(buf), 0)) >= 0)) {
+                mac_ptr = NULL;
+                for (nl_reply_hdr = (struct nlmsghdr *) buf; (! done) && NLMSG_OK(nl_reply_hdr, buff_fill); nl_reply_hdr = NLMSG_NEXT(nl_reply_hdr, buff_fill)) {
+                    if (nl_reply_hdr->nlmsg_type == NLMSG_DONE) done = 1;
+                    else if (nl_reply_hdr->nlmsg_type == RTM_NEWNEIGH) {
+                        nl_reply_msg = (struct nlmsghdr *) NLMSG_DATA(nl_reply_hdr);
+                        rt_reply_len = RTM_PAYLOAD(nl_reply_hdr);
+
+                        for (rt_reply_attr = RTM_RTA(nl_reply_msg); (! done) && RTA_OK(rt_reply_attr, rt_reply_len); rt_reply_attr = RTA_NEXT(rt_reply_attr, rt_reply_len)) {
```

Comment:
Please auto-type in loop iterator. Also the "done" special casing can go when the "return true" is handled as requested.
```suggestion
                        for (auto rt_reply_attr = RTM_RTA(nl_reply_msg); RTA_OK(rt_reply_attr, rt_reply_len); rt_reply_attr = RTA_NEXT(rt_reply_attr, rt_reply_len)) {
```